### PR TITLE
Radiation pulses will yield to MC for every turf they iterate through.

### PR DESCRIPTION
**Pull request by Pickle-Coding**
## About The Pull Request
Changes radiation pulses so that they store the turfs they need to irradiate, iterate through them and remove them from the list of turfs to irradiate, then yield to MC if MC got angry, and put the radiation pulse with the remaining turfs to irradiate to the back of the radiation processing list.

## Why It's Good For The Game
So radiation pulses will listen to the MC more on when to stop.

## Changelog

:cl:
code: Radiation pulses will ask MC if they should continue processing for every turf they iterate through, rather than just every pulse.
/:cl:

### CODE REVIEW

**Comment by Pickle-Coding** [comment]
This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet. Should probably do somethint about that.

**Comment by Pickle-Coding** [comment]
Actually I am not sure if there is a solution that is performant and doesn't have other issues, so I'll just leave it to maintainer discretion on whether I should store and iterate through movables instead of turfs or not.

File: code/__HELPERS/radiation.dm
Comment (LemonInTheDark) [inline code/__HELPERS/radiation.dm:38]: I'd recommend `RANGE_TURFS` here, which uses the internal block() proc, as laid out in the 8 year old profile series done by Rastaf0 <https://gist.github.com/Giacom/be635398926bb463b42a>
It's a bit faster, and also cleaner I think

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:32]: if you're gonna iterate turfs_to_process, you should cache it into a local list variable. I'm not joking it's an unnerving amount faster.
Also rather then using popleft, you could count how many things we've iterated, then cut the list right at the end, either when we're done processing or when we yield. S cleaner and a bit cheaper that way.
Also no proc overhead

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:93]: Rather then reinserting pulse information back into the processing list, why not only remove it from the processing list if pulse() doesn't lead to MC_TICK_CHECK being true? Little cleaner that way

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Good work, some minor comments

**Comment by Mothblocks** [comment]
> This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet

This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

**Comment by Pickle-Coding** [comment]
> > This might actually make a moving object more likely to get irradiated if they move to another turf while the pulse hasn't been complete yet
> 
> 
> 
> This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

I don't think it's likely for someone to get irradiated after they move outside of the radiation source, as it checks for things inside the turfs. The problem with this one is the chance of getting irradiated can be higher than it should be if you're inside the range, if you move into an unprocessed turf while getting irradiated.

**Comment by LemonInTheDark** [comment]
> This is the primary thing I told Lemon to ensure would not happen FYI--if a player moves out of the way of a radiation source, they should not become irradiated a decent time after because of CPU limitations.

The issue here is essentially the inverse, with players moving from a processed turf to an unprocessed turf, or an unprocessed turf to a processed one.
That or moves INTO the path of the wave.

You won't be dosed if you've left the area tho

**Comment by Mothblocks** [comment]
That's a bit crummy but not as bad I think

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:34]: why copy this?

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:99]: make sure to cut here too, since if we yield back we want to remove the turfs we processed

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:101]: why not use the cached thing here?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/radiation.dm
Comment (Pickle-Coding) [reply code/controllers/subsystem/radiation.dm:99]: Doesn't break just break out of the loop, then the cut gets called?

File: code/controllers/subsystem/radiation.dm
Comment (Pickle-Coding) [reply code/controllers/subsystem/radiation.dm:34]: To cache it into a list.

File: code/controllers/subsystem/radiation.dm
Comment (Pickle-Coding) [reply code/controllers/subsystem/radiation.dm:101]: What do you mean by using it here?

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/radiation.dm:34]: It's dumber then that. Just setting a proc var to a datum var caches it and makes access times faster

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/radiation.dm:99]: Oh true true didn't notice that thought it was a return

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [reply code/controllers/subsystem/radiation.dm:101]: See 2 above

File: code/controllers/subsystem/radiation.dm
Comment (Kylerace) [inline code/controllers/subsystem/radiation.dm:37]: guh? whats the point of this just do as anything

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/radiation.dm
Comment (Pickle-Coding) [reply code/controllers/subsystem/radiation.dm]: I don't remember why I added that, but it doesn't serve a purpose so I removed that variable.

File: code/controllers/subsystem/radiation.dm
Comment (LemonInTheDark) [inline code/controllers/subsystem/radiation.dm:37]: It is worth noting that iterations in this form make a copy of the list, so they are not preferable if iterating and yielding over a LARGE list
This is not that tho, by large I mean like, 20k size. See the garbage subsystem.
That is why you'll see the while (thing.length) thing[thing.length] pattern so much though

**Pull request merged**
